### PR TITLE
Adding a helper to merge two colors with different opacities.

### PR DESCRIPTION
**Pull request by mohammadcazig**
Having Multiple layers with different opacity in our components can cause many complexities such as:
Inserting new views,
New API for color mappings,
Difficult to test accessibility,
and etc.

### CODE REVIEW

**Review by jverkoey — CHANGES_REQUESTED** [review]
Please add at least a couple tests verifying black on white and white on black behavior with 100%, 25%, and 0% opacity.

File: components/schemes/Color/src/MDCSemanticColorScheme.m [hunk 1]
```diff
@@ -49,5 +49,19 @@ - (instancetype)initWithDefaults:(MDCColorSchemeDefaults)defaults {
   return self;
 }
 
++ (UIColor *)mergeColor:(UIColor *)color  withBackgroundColor:(UIColor *)backgroundColor {
+  CGFloat red = 0.0, green = 0.0, blue = 0.0, alpha = 0.0;
+  [color getRed:&red green:&green blue:&blue alpha:&alpha];
+  CGFloat bRed = 0.0, bGreen = 0.0, bBlue = 0.0, bAlpha = 0.0;
+  [backgroundColor getRed:&bRed green:&bGreen blue:&bBlue alpha:&bAlpha];
+
+  NSAssert(bAlpha == 1, @"Background color with opacity is not supported.");
```
Comment (ianegordon) [inline components/schemes/Color/src/MDCSemanticColorScheme.m]: Please use a floating point equality check.
Why can't we support non-opaque background colors?
If this is a requirement, please note in the header.
In release code, this assert will be skipped, what is the fallback handling?

File: components/schemes/Color/src/MDCSemanticColorScheme.h [hunk 2]
```diff
@@ -116,4 +116,8 @@ typedef NS_ENUM(NSInteger, MDCColorSchemeDefaults) {
  */
 - (nonnull instancetype)initWithDefaults:(MDCColorSchemeDefaults)defaults;
 
+/** RGBA to RGB conversion */
```
Comment (ianegordon) [inline components/schemes/Color/src/MDCSemanticColorScheme.h]: Naming is hard.
blendColor:withBackgroundColor: may be more specific.
Please give a detailed description on how / why the two colors are merged.
(RGBA -> RGB is super detailed)

Is this the equivalent of blendOverlayMode?
https://developer.apple.com/documentation/coregraphics/cgblendmode/kcgblendmodeoverlay?language=objc

**Review by ianegordon — DISMISSED** [review]

File: components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m [hunk 3]
```diff
@@ -67,4 +67,50 @@ - (void)testInitWithMaterialDefaults {
...
+      [MDCSemanticColorScheme mergeColor:blendColor withBackgroundColor:backgroundColor];
+  UIColor *expectedColor =
+      [UIColor colorWithRed:240.0f/255.0f green:240.0f/255.0f blue:240.0f/255.0f alpha:1.0f];
+
+  XCTAssertTrue([self compareColorsWithFloatPrecisionFirstColor:resultColor
+                                                    secondColor:expectedColor]);
+
+}
+
+- (BOOL)compareColorsWithFloatPrecisionFirstColor:(UIColor *)firstColor
```
Comment (jverkoey) [inline components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m]: Add similar tests for blending on to black as well please.

**Review by jverkoey — DISMISSED** [review]

File: components/schemes/Color/src/MDCSemanticColorScheme.h (hunk 2 above)
Comment (mohammadcazig) [reply components/schemes/Color/src/MDCSemanticColorScheme.h]: I added documentation.
https://en.wikipedia.org/wiki/Alpha_compositing
it's different than blending images.

File: components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m (hunk 3 above)
Comment (mohammadcazig) [reply components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m]: Done.

**Review by jverkoey — APPROVED** [review]

**Comment by jverkoey** [comment]
Failures appear to be legitimate:

```
ERROR: /Volumes/BuildData/tmpfs/src/github/repo/components/schemes/Color/BUILD:24:1: C++ compilation of rule '//components/schemes/Color:Color' failed (Exit 1)
components/schemes/Color/src/MDCSemanticColorScheme.m:58:17: error: incompatible pointer types sending 'double *' to parameter of type 'CGFloat * _Nullable' (aka 'float *') [-Werror,-Wincompatible-pointer-types]
  [color getRed:&red green:&green blue:&blue alpha:&alpha];
                ^~~~
```

**Review by ianegordon — DISMISSED** [review]

**Review by ianegordon — DISMISSED** [review]
Code LGTM, but please fix conversion errors and consider changing parameters from double to CGFloat to match the values.

File: components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m [hunk 4]
```diff
@@ -67,4 +67,58 @@ - (void)testInitWithMaterialDefaults {
...
+- (void)testColorMergeFor50OpacityWhiteOnBlack {
+  UIColor *backgroundColor = [UIColor blackColor];
+  UIColor *blendColor = [UIColor colorWithRed:1 green:1 blue:1 alpha:0.5];
+  UIColor *expectedColor =
+      [UIColor colorWithRed:0.5f green:0.5f blue:0.5f alpha:1.0f];
+  UIColor *resultColor =
+      [MDCSemanticColorScheme blendColor:blendColor withBackgroundColor:backgroundColor];
+  XCTAssertEqualObjects(resultColor, expectedColor);
+}
+- (BOOL)compareColorsWithFloatPrecisionFirstColor:(UIColor *)firstColor
```
Comment (romoore) [inline components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m]: Please add tests for HSB (including beyond 1.0), P3, and greyscale color spaces.

File: components/schemes/Color/src/MDCSemanticColorScheme.m [hunk 5]
```diff
@@ -22,6 +22,10 @@
                           blue:(CGFloat)((colorValue & 0xFF) / 255.0) alpha:1];
 }
 
+static double blendColorChannel(double value, double bValue, double alpha, double bAlpha) {
```
Comment (romoore) [inline components/schemes/Color/src/MDCSemanticColorScheme.m]: Please add documentation for parameters.

File: components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m [hunk 6]
```diff
@@ -67,4 +67,58 @@ - (void)testInitWithMaterialDefaults {
...
+}
+
+- (void)testColorMergeFor60GrayOpacityOnWhite {
+  UIColor *backgroundColor = [UIColor whiteColor];
+  UIColor *blendColor =
+      [UIColor colorWithRed:230.0f/255.0f green:230.0f/255.0f blue:230.0f/255.0f alpha:0.60f];
+  UIColor *resultColor =
+      [MDCSemanticColorScheme blendColor:blendColor withBackgroundColor:backgroundColor];
+  UIColor *expectedColor =
+      [UIColor colorWithRed:240.0f/255.0f green:240.0f/255.0f blue:240.0f/255.0f alpha:1.0f];
```
Comment (romoore) [inline components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m]: Please add a test with non-uniform color channel information to detect if the color channels get mixed by accident.

File: components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m (hunk 4 above)
Comment (mohammadcazig) [reply components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m]: Added more test for HSB and grey scale for now.

File: components/schemes/Color/src/MDCSemanticColorScheme.m (hunk 5 above)
Comment (mohammadcazig) [reply components/schemes/Color/src/MDCSemanticColorScheme.m]: Done.

File: components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m (hunk 6 above)
Comment (mohammadcazig) [reply components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m]: Done.

File: components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m (hunk 4 above)
Comment (romoore) [reply components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m]: Does this work for DisplayP3?

File: components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m [hunk 7]
```diff
@@ -149,10 +149,11 @@ - (BOOL)compareColorsWithFloatPrecisionFirstColor:(UIColor *)firstColor
   [firstColor getRed:&fRed green:&fGreen blue:&fBlue alpha:&fAlpha];
   CGFloat sRed = 0.0, sGreen = 0.0, sBlue = 0.0, sAlpha = 0.0;
   [secondColor getRed:&sRed green:&sGreen blue:&sBlue alpha:&sAlpha];
-  return (roundf(fRed) == roundf(sRed) &&
-          roundf(fGreen) == roundf(sGreen) &&
-          roundf(fBlue) == roundf(sBlue) &&
-          roundf(fAlpha) == roundf(sAlpha));
+
+  return ((float)fRed == (float)sRed &&
```
Comment (romoore) [inline components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m]: Nit: Please consider using MDCFloatEquals

**Review by romoore — APPROVED** [review]
If not in this PR, please create a backlog issue to verify display P3 color space support.

**Pull request merged**